File: bindings/react-native/ios/StorageNamespace.swift
//
// StorageNamespace.swift
// OfflineProtocol
//
// Stable account namespaces for built-in protocol storage.
//

import CryptoKit
import Foundation

enum StorageNamespace {
    private static let domain = "offline-protocol-storage-v1"
    private static let hex = Array("0123456789abcdef".utf8)

    /// A namespace becomes a path component and a Keychain service suffix, so
    /// an arbitrary caller-supplied string could escape its own account (`..`)
    /// or collide with another's. Mirrors `requireAccount` on Android and
    /// `require_account_storage_namespace` in Python.
    static func requireAccount(_ value: String) throws -> String {
        let digits = value.dropFirst("account-".count)
        guard value.hasPrefix("account-"),
              digits.count == 64,
              digits.allSatisfy({ $0.isHexDigit && !$0.isUppercase })
        else {
            throw MlsStorageError.StoreFailed(
                message: "Invalid protocol storage account namespace"
            )
        }
        return value
    }

    static func account(appId: String, userId: String) -> String {
        let input = "\(domain)\0\(appId)\0\(userId)"
        let digest = SHA256.hash(data: Data(input.utf8))
        var output = Array("account-".utf8)
        output.reserveCapacity(output.count + SHA256.byteCount * 2)
        for byte in digest {
            output.append(hex[Int(byte >> 4)])
            output.append(hex[Int(byte & 0x0f)])
        }
        return String(decoding: output, as: UTF8.self)
    }
}

File: bindings/python/offline_protocol_sdk/storage_namespace.py
"""Stable account namespaces for built-in protocol storage."""

from __future__ import annotations

import hashlib
import re

_NAMESPACE_DOMAIN = b"offline-protocol-storage-v1"

_ACCOUNT_PATTERN = re.compile(r"\Aaccount-[0-9a-f]{64}\Z")


def account_storage_namespace(app_id: str, profile: str) -> str:
    """Return an opaque, filesystem-safe namespace for one protocol account."""

    digest = hashlib.sha256(
        _NAMESPACE_DOMAIN + b"\0" + app_id.encode("utf-8") + b"\0" + profile.encode("utf-8")
    ).hexdigest()
    return f"account-{digest}"


def require_account_storage_namespace(value: str) -> str:
    """Return *value* if it is a well-formed account namespace, else raise.

    Built-in stores turn a namespace into a path or credential-store component,
    so an arbitrary caller-supplied string could escape its own account (``..``)
    or collide with another's. Mirrors ``StorageNamespace.requireAccount`` on
    Android and iOS.
    """

    if not _ACCOUNT_PATTERN.match(value):
        raise ValueError("Invalid protocol storage account namespace")
    return value
